packages/language-service/test/completions_spec.ts

```
@@ -479,7 +479,7 @@ describe('completions', () => {
       }
     });
 
-    it('should be able to resolve lowercase', () => {
+    it('should be able to resolve json', () => {
```

As pointed out in the commit message, this seems wrong. I need some guidance to fix this properly, as the error message looks pretty generic (and not very readable, sorry):
```
Expected [ Object({ name: 'ng-container', kind: 'angular element', sortText: 'ng-container', replacementSpan: Object({ start: 760, length: 9 }) }), Object({ name: 'ng-content', kind: 'angular element', sortText: 'ng-content', replacementSpan: Object({ start: 760, length: 9 }) }), Object({ name: 'ng-template', kind: 'angular element', sortText: 'ng-template', replacementSpan: Object({ start: 760, length: 9 }) }), Object({ name: 'a', kind: 'html element', sortText: 'a', replacementSpan: Object({ start: 760, length: 9 }) }), Object({ name: 'abbr', kind: 'html element', sortText: 'abbr', replacementSpan: Object({ start: 760, length: 9 }) }), Object({ name: 'acronym', kind: 'html element', sortText: 'acronym', replacementSpan: Object({ start: 760, length: 9 }) }), Object({ name: 'address', kind: 'html element', sortText: 'address', replacementSpan: Object({ start: 760, length: 9 }) }), Object({ name: 'area', kind: 'html element', sortText: 'area', replacementSpan: Object({ start: 760, length: 9 }) ... to contain <jasmine.objectContaining(Object({ name: 'substring', kind: 'method' }))>.
```

I believe this might be related to an issue in resolving the type of the pipe when multiple overloads are available.

---

Yep, the language service does not take argument types into account when determining the type of the result. Apparently it can only:
 - resolve the type for pipes with no overload
 - resolve the type for `async` and `slice`, which are [hardcoded](https://github.com/angular/angular/blob/master/packages/language-service/src/typescript_symbols.ts#L622-L644) (possibly in an incorrect way... it looks like the `async` is hardcoded as if it the pipe never returned `null`)

---

@kyliau do you have any thoughts here? Should we wait for the new integration with ngtsc?

---

This has been on my radar for a while. I'd love to spend some cycles this week to see if there's anything we can do to improve the logic for picking function signatures. Integration with ngtsc looks like it's gonna take a while.
@ranma42 Could you please rebase? A lot have changed since your last commit, and this test case in particular has been refactored quite a bit.

---

I rebased the branch, which is currently running on the CI
Before merging I would love to improve the last commit, but if there are no short-term plans for completing the integration with ngtsc we might prefer to merge it (and possibly add a "should-fail" test that uses lowercase and that will "unexpectedly" start passing when the integration is completed).

I do not know this test system very well, so I do not know if/how such tests are done. Should I just test the opposite and have a comment explaining that the test is bad but reflects the current state of the codebase? or is there a specific syntax for well-known failures?

---

This is resolved in https://github.com/kyliau/angular/commit/3f64f0b0b3f6e03d38d10e64d175b94cbd41a48f